src/main/java/com/minecolonies/structures/helpers/Structure.java

```diff
@@ -82,63 +91,361 @@
+            return null;
+        }
+        return calculateMD5(getStreamAsByteArray(stream));
+    }
+
+    /**
+     * Calculate the MD5 hash of a byte array
+     *
+     * @param bytes array
+     * @return the MD5 hash string or null
+     */
+    public static String calculateMD5(final byte[] bytes) //throws IOException
+    {
         try
         {
-            this.template = readTemplateFromStream(inputstream);
+            final MessageDigest md = MessageDigest.getInstance("MD5");
+            return DatatypeConverter.printHexBinary(md.digest(bytes));
         }
-        catch (final IOException e)
+        catch (@NotNull NoSuchAlgorithmException e)
         {
-            Log.getLogger().warn(String.format("Failed to load template %s", structureName), e);
+            Log.getLogger().trace(e);
         }
-        finally
+
+        return null;
+    }
+
+    /**
+     * Compare the md5 from the structure with an other md5 hash.
+     *
+     * @param otherMD5 to compare with
+     * @return whether the otherMD5 match, return false if md5 is null
+     */
+    public boolean isCorrectMD5(final String otherMD5)
+    {
+        Log.getLogger().info("isCorrectMD5: md5:" + md5 + " other:" + otherMD5);
+        if (md5 == null || otherMD5 == null)
         {
-            IOUtils.closeQuietly(inputstream);
+            return false;
         }
+        return md5.compareTo(otherMD5) == 0;
+    }
+
+    public static byte[] compress(final byte[] data)
+    {
+        try
+        {
+            final ByteArrayOutputStream byteStream = new ByteArrayOutputStream(data.length);
+            try
+            {
+                try (GZIPOutputStream zipStream = new GZIPOutputStream(byteStream))
+                {
+                    zipStream.write(data);
+                }
+            }
+            finally
+            {
+                byteStream.close();
+            }
+
+            return byteStream.toByteArray();
+        }
+        catch (Exception e)
+        {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    public static byte[] uncompress(final byte[] data)
+    {
+        byte[] buffer = new byte[1024];
+        final ByteArrayOutputStream out = new ByteArrayOutputStream();
+
+        try
+        {
+            try (ByteArrayInputStream byteStream = new ByteArrayInputStream(data))
+            {
+                try (GZIPInputStream zipStream = new GZIPInputStream(byteStream))
+                {
+                    int len;
+                    while ((len = zipStream.read(buffer)) > 0)
+                    {
+                        out.write(buffer, 0, len);
+                    }
+                }
+            }
+
+            return out.toByteArray();
+        }
+        catch (Exception e)
+        {
+            Log.getLogger().warn("Could not uncompress data:" + e.getMessage());
+        }
+        return null;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Return an empty array instead of null. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1168)